ModuleThreeFirstTaskConsole/FileSystemVisitor.cs

```diff
@@ -0,0 +1,196 @@
+            Func<IFileSystemInfo, bool> filter,
+            string initialPath)
+        {
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _filter = filter is null ? x => true : filter;
+            _initialPath = initialPath ?? STANDARDPATH;
+            _initialPath = _fileSystem.Directory.Exists(initialPath) ? initialPath : throw new DirectoryNotFoundException("You entered path that doesn't exist.");
+            Stoped = true;
+        }
+
+        /// <summary>
+        /// Events that happen at the begin of search.
+        /// </summary>
+        public event EventHandler<IFileSystemInfo> SearchStarted;
+
+        /// <summary>
+        /// Events that happen at the end of search.
+        /// </summary>
+        public event EventHandler<IFileSystemInfo> SearchEnded;
+
+        /// <summary>
+        /// Events that happen when file is found.
+        /// </summary>
+        public event EventHandler<FileSystemEventArgs> FileFound;
+
+        /// <summary>
+        /// Events that happen when directory is found.
+        /// </summary>
+        public event EventHandler<FileSystemEventArgs> DirectoryFound;
+
+        /// <summary>
+        /// Events that happen when filtered file is found.
+        /// </summary>
+        public event EventHandler<FileSystemEventArgs> FilteredFileFound;
+
+        /// <summary>
+        /// Events that happen when filtered directory is found.
+        /// </summary>
+        public event EventHandler<FileSystemEventArgs> FilteredDirectoryFound;
+
+        /// <summary>
+        /// State of searching. If it's true on the next step except for search start the visitor will not proceed to work.
+        /// </summary>
+        public bool Stoped { get; set; }
+
+        /// <summary>
+        /// Search files in provided fileSystem.
+        /// </summary>
+        /// <returns>Files from directories.</returns>
+        public async IAsyncEnumerable<string> Search()
+        {
+            var initialDiractoryName = _fileSystem.Path.GetDirectoryName(_initialPath);
+            var directory = _fileSystem.DirectoryInfo.FromDirectoryName(initialDiractoryName ?? _fileSystem.Directory.GetCurrentDirectory());
+            OnSearchStarted(directory);
+            await foreach (var info in GetAllFromCurrentDirectory(directory))
+            {
+                yield return GetNameWithoutLongPath(info);
+                if (Stoped)
+                {
+                    break;
+                }
+            }
+
+            OnSearchEnded(directory);
+        }
+
+        /// <summary>
+        /// Returns every single directory and filtered files from sprecified directory.
+        /// </summary>
+        /// <param name="directory">Provided directory.</param>
+        /// <returns>FileInfo of each file and directory.</returns>
+        private async IAsyncEnumerable<IFileSystemInfo> GetAllFromCurrentDirectory(IFileSystemInfo directory)
```

Comment:
asynchronous stream is still useless here

Reply:
Fixed in a883af4